src/directives/repeat.ts

```diff
@@ -41,67 +96,139 @@ export function repeat<T>(
     keyFn = keyFnOrTemplate as KeyFn<T>;
   }
 
-  return directive((part: NodePart): void => {
-    let keyMap = keyMapCache.get(part);
-    if (keyMap === undefined) {
-      keyMap = new Map();
-      keyMapCache.set(part, keyMap);
-    }
-    const container = part.startNode.parentNode as HTMLElement | ShadowRoot |
-        DocumentFragment;
-    let index = -1;
-    let currentMarker = part.startNode.nextSibling!;
+  return directive((directivePart: NodePart): void => {
+    // Old part list & map is retrieved from the last render (associated with
+    // the part created for this instance of the directive)
+    let oldParts = partListCache.get(directivePart) || [];
+    let oldKeyToIndexMap = partMapCache.get(directivePart) || new Map();
 
-    for (const item of items) {
-      let result;
-      let key;
-      try {
-        ++index;
-        result = template !(item, index);
-        key = keyFn ? keyFn(item) : index;
-      } catch (e) {
-        console.error(e);
-        continue;
-      }
+    // New part list will be built up as we go (either reused from old parts or
+    // created for new keys in this update). This is saved in partListCache
+    // at the end of the update.
+    const newParts: Array<KeyedNodePart> = [];
+    const newKeyToIndexMap: Map<any, number> = new Map();
 
-      // Try to reuse a part
-      let itemPart = keyMap.get(key);
-      if (itemPart === undefined) {
-        // TODO(justinfagnani): We really want to avoid manual marker creation
-        // here and instead use something like part.insertBeforePart(). This
-        // requires a little refactoring, like iterating through values and
-        // existing parts like NodePart#_setIterable does. We can also remove
-        // keyMapCache and use part._value instead.
-        // But... repeat() is badly in need of rewriting, so we'll do this for
-        // now and revisit soon.
-        const marker = createMarker();
-        const endNode = createMarker();
-        container.insertBefore(marker, currentMarker);
-        container.insertBefore(endNode, currentMarker);
-        itemPart = new NodePart(part.templateFactory);
-        itemPart.insertAfterNode(marker);
-        if (key !== undefined) {
-          keyMap.set(key, itemPart);
-        }
-      } else if (currentMarker !== itemPart.startNode) {
-        // Existing part in the wrong position
-        const end = itemPart.endNode.nextSibling!;
-        if (currentMarker !== end) {
-          reparentNodes(container, itemPart.startNode, end, currentMarker);
-        }
+    // New result list is eagerly generated from items and marked with its key.
+    // This is saved in partMapCache at the end of the update.
+    const newResults: Array<KeyedTemplateResult> = [];
+    let index = 0;
+    for (let item of items) {
+      const key = keyFn ? keyFn(item) : index;
```

Comment:
This was what you [initially had](https://github.com/Polymer/lit-html/blob/7bb21437d32a5f8d95f977e8659b2627bb7fbaf0/src/directives/repeat.ts#L61), which I assumed was intentional to have the non-keyed mode just work like map. 

If we switched to defaulting to use the item, it might be slightly counterintuitive for the typical immutable pattern case, since any changed items would have their DOM discarded and re-created, since that's the current contract with keys is to never update a part with a different key (at least until we added a pooling mode that reuses parts). WDYT?

Reply:
Now I'm questioning my original decision :) but at least this way doesn't result in any moves, and we hope that updating a homogeneous list of TemplateResults is fast... Like you said, just like .map()

Reply:
I just want to make sure we're clear; I'm maintaining the `key` contract as follows:
* "always update this part with results from this key" AND 
* "never update this part with results from a different key"

Which means that in the typical immutable pattern e.g. 
```js
items = [...items.slice(0,3), {...items[3], selected: true}, ...items.slice(4)]
```
...the part at index 3 and all of its DOM will be thrown away and re-created because its key (the object) changed.  And worse, "select all" would cause all the DOM to be thrown out.  Seems more footgunny than keeping it with `index`.  But it's arguable. Since map will always be available, that would provides a 3rd option I suppose.  But really, who's savvy enough to use `repeat` and not provide a key? 🤷‍♂️.  @sorvell WDYT?

Reply:
Oh but like I said above, _if_ we added a mode to allow reusing removed parts with new parts it would make more sense. Because the operation above would result in 1 part being pooled (the old object no longer in the array) and 1 part being added from the pool (the new object which replaced it), which would come out a wash and also avoid moves, without providing a key.  But I'm holding off adding that in this PR.

I sorta feel like it would be better right now to not have the varargs signature and require a key function, which gives us flexibility to decide on the non-keyed behavior later.

Reply:
I am sure that people coming from Polymer land might very well use repeat as a replacement for map and thus without a key. I believe it is quite important to explain repeat well in the documentation, like when to use and when no to, etc